Ml 357 GLM bundle creation

timothyesler opened the pull request:
@johnholt, as with the SVM bundle, I have commited the existing code that I copied from the LogisticRegression bundle as an initial commit, and my changes as a second commit.

Commit:
ML-357 GLM Bundle

Exact copy of LogisticRegression (starting point).

Signed-off-by: Timothy Esler <timothy.esler@lexisnexisrisk.com>

Commit:
ML-357 GLM Bundle

Extension of the logistic regression bundle to generalized linear models in general. This included:
- Generalization of IRLS algorithm for all GLM distribution families
- Generalization of deviance and log-likelihood calculations
- Specification of link, mean, initialization, variance, log-likelihood, etc. function for binomial, quasibinomial, Poisson, quasipoisson, gamma, Gaussian, and inverse Gaussian error distributions
- Addition of a weights argument
- Single model training and prediction interface
- Addition of data distribution validation
- Addition of unit tests for all error distributions and for data validation messages

Signed-off-by: Timothy Esler <timothy.esler@lexisnexisrisk.com>

johnholt commented on Constants.ecl:
```diff
@@ -1,19 +1,19 @@
-// Constants used by Logistic Regression.  Most of these are the
+﻿// Constants used by Logistic Regression.  Most of these are the
```
Didn't you mean to change the name to GLM here?

johnholt commented on Constants.ecl:
```diff
@@ -1,19 +1,19 @@
   EXPORT REAL8 default_epsilon := 0.00000001;
   EXPORT REAL8 default_ridge := 0.00001;
-  EXPORT UNSIGNED4 local_cap := 1000000 : STORED('LR_LOCAL_MATRIX_CAP');
+  EXPORT UNSIGNED4 local_cap := 5000000 : STORED('LR_LOCAL_MATRIX_CAP');
```
The name that you use in the test attribute below is GLM_ so this setting should be GLM_LOCAL_MATRIX_CAP.

johnholt commented on DataStats.ecl:
```diff
@@ -1,49 +1,53 @@
   l1_dep := GROUP(dep(field_details), wi, LOCAL);
   l1_dep_srt := SORT(l1_dep, number , value);
-  l1_dep_grp := GROUP(l1_dep_srt, number);
+  l1_dep_grp := GROUP(l1_dep_srt, number, LOCAL);
```
This will not always work as you want.  The SORT(...) above will leave the data sequenced and distributed such that every record with the same values for number and value will be on the same node, BUT, not all records for a particular value of number will be on the same node.  By adding LOCAL, you can strand some records on a previous node.   Without LOCAL, the GROUP operation will migrate records up as necessary.

timothyesler replied:
@johnholt could you tell me where I'm going wrong, I'd like to see if there's a way to get a similar speedup in a similar way if this approach doesn't work. My thinking was that:

```l1_dep := GROUP(dep(field_details), wi, LOCAL);``` will put all records with the same wi on the same node, then ```l1_dep_srt := SORT(l1_dep, number , value);``` will keep that grouping and distribution (as it's implicitly local) but will sort within each grouping by value and number. Finally, ```l1_dep_grp := GROUP(l1_dep_srt, number, LOCAL)``` would leave the dataset grouped by wi and number (and sorted by value within each group). So that there may be records with the same number on different nodes, but only if they are from different work items, which is okay.

Let me know where the mistake is and try find another solution.

johnholt replied:
I think I missed the earlier I1_dep definition that groups the data by wi locally.  My comment is incorrect.

However, the original sequence is not what you describe.  I1_dep does not move any records across nodes or even within the nodes.  So, I1_dep can have more than one node with records from the same work_item, AND, there can be more than 1 group for each work_item value on the same node.

Since I1_dep is already grouped, the LOCAL flag on I1_dep_grp does nothing.  I think that it should have generated a warning though.

The idea in the original sequence was to perform a local reduction to unique strings, and then if there were not too many instances determine a global list of unique strings.

johnholt commented on Family.ecl:
```diff
@@ -0,0 +1,128 @@
+    EXPORT REAL8 var (REAL8 m);
+    EXPORT REAL8 init(REAL8 y, REAL8 w);
+    EXPORT REAL8 ll  (REAL8 y, REAL8 m, REAL8 disp);
+    EXPORT VARSTRING mu_LUCI;
```
STRING should be more efficient.  VARSTRING is a good choice when working with native C/C++ functions and external data.

johnholt commented on Family.ecl:
```diff
@@ -0,0 +1,128 @@
+  EXPORT FamilyInterface := INTERFACE
+    EXPORT REAL8 link(REAL8 m);
+    EXPORT REAL8 mu  (REAL8 v);
+    EXPORT REAL8 deta(REAL8 m);
```
deta seems like a strange choice for a public interface name.  Does the name have any significance?

johnholt commented on IRLS/GetModel_global.ecl:
```diff
@@ -36,92 +128,65 @@ dimm         := LR.dimm;
+                  dimension_t c) := w_i(u) * POWER(res, 2);
+
+  // check for user-defined weights
+  doWeights := COUNT(weights) > 0;
```
doWeights := EXISTS(weights); is a better (cheaper) test.

johnholt commented on IRLS/GetModel_global.ecl:
```diff
@@ -285,12 +357,13 @@ EXPORT DATASET(Layout_Model)
     SELF.frst_addr := inf.frst_addr;
     SELF.parts := inf.parts;
-    SELF.mat := make_vector(inf.dims, 0.00000001); // make non-zero
+    SELF.mat := make_vector(inf.dims, 0.0); // make non-zero
```
The comments should match what you want to do.

johnholt commented on IRLS/GetModel_local.ecl:
```diff
@@ -210,7 +261,7 @@ EXPORT DATASET(Layout_Model)
   // Initial beta estimate
   Part init_beta(Part x, Part y) := TRANSFORM
-    SELF.mat := make_vector(x.part_cols, 0.00000001); // make non-zero
+    SELF.mat := make_vector(x.part_cols, 0.0); // make non-zero
```
Fix comment to match or drop the comment.

johnholt reviewed (changes requested):
Only a few problems to fix.

Commit:
ML-357 GLM Bundle

Changes in response to review of pull request:
- Removed references in LR
- More control of what DataStats does, to enable faster computation
- Added explanatory comments to Family.ecl interface
- Fixed some comments

Validation and performance tests have been rerun.

Signed-off-by: Timothy Esler <timothy.esler@lexisnexisrisk.com>

ghost commented:
Commit added to address the above comments. I've attached the results of the LR vs GLM performance tests. The majority (perhaps all?) of the added time comes from the data validation performed by GLM.
[GLMvsLR.xlsx](https://github.com/hpcc-systems/GLM/files/1342836/GLMvsLR.xlsx)
